docs/langref/relay/type_system.rst

```diff
@@ -0,0 +1,187 @@
+===========
+
+We have briefly introduced types while detailing the the expression language
+of Relay, but have fully laid out the type system.
+
+Although the majority of Relay programs are written without type annotations,
+Relay is statically typed.
+
+Static types are useful because they enable efficient layout, memory reuse, and
+code generation. They aid in debugging program transformations, but can also
+give us the expressivity afforded by more dynamic langauges.
+
+We are able to omit these type annotations by a process known as type inference.
+Type inference is a technique that has its roots in the programming language
+community, and can be viewed as a method for generalizing shape inference to
+run over arbitrary user programs containing control flow and recursion.
+
+Static types are useful when performing compiler optimization because they
+communicate properties about the data we manipulate, such as runtime shape,
+data layout, storage without needing to run the program.
+
+Most current IRs use "shape inference" to recover Tensor dimensions from the user
+provided program. Machine learning users have enjoyed shape inference for
+tensors because it allows them to generate performant code without giving up
+on the expressivity of the input language.
+
+Because Relay is intended as an IR we require *some* type information to provide
+full inference. We don't believe this to be an issue as many of the IR builder
+inferfaces require some type information, or can generate IR based on their own
+higher level inferences.
+
+We view this limited shape inference as a simpler form of type
+inference. Instead of relying on an ad-hoc procedure for recovering type
+information from a potentially dynamic program, we apply ideas from compiler and IR design.
+
+Below we briefly dicsuss the different kinds of types in Relay.
+
+=====
+Types
+=====
+
+Relay's type system has a "language of types" which allow us to write down the type of
+a Relay program. Below we detail the langauge of types and how we assign them to Relay
+programs.
+
+Type
+~~~~
+The base type for all Relay types. All Relay types are sub-classes of this base type.
+
+See :py:class:`~tvm.relay.type.Type` for its definition and documentation.
+
+Tensor Type
+~~~~~~~~~~~
+
+A concrete TensorType in Relay, see tvm/relay/type.h for more details.
+
+This is the type assigned to tensor's with a known dype and shape. For
+example a tensor of `float32` and `(5, 5)`.
+
+
+
+Tensor values in Relay are typed with tensor types. A tensor type is
+parametrized by a data type, and shape. The data type must be a base
+type as enforced by the kind checking rules described in TODO.
+
+This restriction importantly means
+
+The shape may be any valid Relay shape as described in the below
+section on shapes.
+
+
+See :py:class:`~tvm.relay.type.TensorType` for its definition and documentation.
+
+Kind
+~~~~
+The kind of a type parameter, represents a variable shape,
+base type, type, or dimension.
```

Comment:
what's the difference between `base type` and `type`? Actually, a little confused about this part. Can I understand `Kind` as type's type?

Reply:
Kind is basically a type's type.
base type is type that has a simple data layout, and thus can act as element of tensor efficiently.